src/graph_synchronizer.cpp

```diff
@@ -122,37 +119,36 @@ void GraphSynchronizer::Lock::lock() {
             // Make them into pos_ts that point left to right, the way Jordan thinks.
             pos_t left_pos = make_pos_t(start_left.node, start_left.is_end, 0);
             pos_t right_pos = make_pos_t(end_right.node, !end_right.is_end,
-                synchronizer.graph.get_node(end_right.node)->sequence().size() - 1);
+                synchronizer.graph.get_node(end_right.node)->sequence().size());
             
             // Since these are already at node ends, we don't need to worry about node cuts.
             
             // Extract paths out to the length we need to connect the ends, or a bit further.
             // TODO: be sure to extract really big indels somehow...
-            auto duplications = algorithms::extract_connecting_graph(&synchronizer.graph,
-                context_graph,
+            auto translator = algorithms::extract_connecting_graph(&synchronizer.graph,
+                &context,
                 (past_end - start) * 2,
                 left_pos,
                 right_pos,
-                true, // Keep the specified positions in the graph
                 false, // Disallow terminal node cycles, so we don't duplicate nodes
-                true, // Remove other tips
-                false, // We don't care about being on a path
-                false); // Or being strictly less than the specified length
+                true, // We don't want extraneous material that doesn't connect the positions
+                false); // But we don't care about being strictly less than the specified length
                 
 #ifdef debug
             cerr << "Extracted " << context_graph.node_size() << " nodes and " << context_graph.edge_size() << " edges between " << path_name << ":" << start << "-" << past_end << endl;
 #endif
                 
-            // Any duplication is going to mess things up, since we need
+            // Any ID mismatch is going to mess things up, since we need
             // operations on the new graph to make sense in the original graph.
-            // We need all the entries in this duplication map to be no-ops.
-            for (auto& kv : duplications) {
-                assert(kv.first == kv.second);
+            // We need all the entries in this translation map to be no-ops, so
+            // we translate all IDs back to their original (which is possible
+            // because we chose extraction paramters that never duplicate nodes).
```

Comment:
The documentation for extract_connecting_graph doesn't seem to specify when nodes may be duplicated. If we rely on the lack of duplication here, we ought to put it in the function spec that under these particular conditions there will be no duplication. This new swap code won't actually detect and abort if duplication *has* occurred when it wasn't supposed to.